File: Recursion/lecture 7/Fraz_recursion_7_2.cpp

void helper(vector<int> &v,int i, vector<int> &subSet,vector<vector<int>> &ans)
{
    if(i==v.size())
    {
        ans.push_back(subSet);
        return;
    }
    //include the ith element
    subSet.push_back(v[i]);
    helper(v,i+1,subSet,ans);
    // not including the ith element
    subSet.pop_back();
    helper(v,i+1,subSet,ans);
}

vector<vector<int>> pwset(vector<int>v)
{
    vector<vector<int>> ans;
    vector<int> subSet;
     helper(v,0,subSet,ans);
    return ans;
}

File: Recursion/lecture 7/Fraz_recursion_7_2.java
ArrayList<ArrayList<Integer>> ans;
ArrayList<Integer> subSet;

void helper(ArrayList<Integer> v, int i, ArrayList<Integer> subSet, ArrayList<ArrayList<Integer>> ans) {
        if (i == v.size()) {
            ans.add(new ArrayList(subSet));
            return;
        }
        // include the ith element
        subSet.add(v.get(i));
        helper(v, i + 1, subSet, ans);
        // not including the ith element
        subSet.remove(subSet.size()-1);
        helper(v, i + 1, subSet, ans);
}

ArrayList<ArrayList<Integer>> pwset(ArrayList<Integer> v) {
        helper(v, 0, subSet, ans);
        return ans;
}
